File: go/linked_list/linked_list.go
package linked_list

type LinkedList struct {
	head   *Node
	tail   *Node
	length int
}

type Node struct {
	Value int
	Next  *Node
}

func (l *LinkedList) Empty() bool {
	if l.head == nil {
		return true
	}
	return false
}

func (l *LinkedList) Head() *Node {
	return l.head
}

func (l *LinkedList) Tail() *Node {
	return l.tail
}

func (l *LinkedList) Lenght() int {
	return l.length
}

func (l *LinkedList) InsertBefore(in *Node, before *Node) *Node {
	if !l.Empty() {
		n, prev := l.head, l.head
		for n != nil {
			if n.Value == before.Value {
				if n == l.head {
					return l.Push(in)
				}

				prev.Next = in
				in.Next = n
				l.length++

				return in
			}
			prev = n
			n = n.Next
		}
		return nil
	}
	return nil
}

func (l *LinkedList) InsertAfter(in *Node, after *Node) *Node {
	if !l.Empty() {
		n := l.head
		for n != nil {
			if n.Value == after.Value {
				if n == l.tail {
					return l.InsertLast(in)
				}
				in.Next = n.Next
				n.Next = in
				l.length++

				return in
			}
			n = n.Next
		}
		return nil
	}
	return nil
}

func (l *LinkedList) InsertLast(n *Node) *Node {
	if l.Empty() {
		return l.Push(n)
	}
	l.tail.Next = n
	l.tail = n
	l.length++
	return n
}

func (l *LinkedList) Reverse() {
	n := l.head.Next
	prev := l.head
	for n != nil {
		next := n.Next
		n.Next = prev
		prev = n
		n = next
	}
	l.head, l.tail = l.tail, l.head
	l.tail.Next = nil
}

func (l *LinkedList) Remove(s *Node) *Node {
	if !l.Empty() {
		n, prev := l.head, l.head
		for n != nil {
			if n.Value == s.Value {
				l.length--
				if n == l.head {
					l.head = n.Next
					return n
				}
				if n == l.tail {
					l.tail = prev
					l.tail.Next = nil
					return n
				}
				prev.Next = n.Next
				return n
			}
			prev = n
			n = n.Next
		}
		return nil
	}
	return nil
}

func (l *LinkedList) Push(n *Node) *Node {
	if l.Empty() {
		l.head, l.tail = n, n
	} else {
		prev := l.head
		n.Next = prev
		l.head = n
	}
	l.length++
	return n
}

func (l *LinkedList) Pop() *Node {
	if !l.Empty() {
		n := l.head
		l.head = l.head.Next
		l.length--

		return n
	}
	return nil
}

File: ruby/linked_list/LinkedList.rb
class LinkedList
  def initialize(head = nil)
    @head = head
    if head
      @count = 1;
    else
      @count = 0;
    end
  end

  def getCount
    @count
  end

  def print_list
    _current = @head
    unless _current
      puts "List is empty!"
    else
      while _current do
        puts "Node: #{_current.getValue}"
        _current = _current.getNext
      end
    end
  end

  def insert(node)
    unless @head
      @head = node
      @count += 1
    else
      _current = @head
      while _current.getNext do
        _current = _current.getNext
      end
      _current.setNext(node)
      @count += 1
    end
  end

  def insert_beginning(node)
    unless @head
      @head = node
      @count += 1
    else
      node.setNext(@head)
      @head = node
      @count += 1
    end
  end

  def insert_after(node, afterNode)
    _current = @head
    if _current.getValue == afterNode.getValue
      node.setNext(_current.getNext)
      _current.setNext(node)
      @count += 1
    else
      while _current && _current.getNext do
        if _current.getNext.getValue == afterNode.getValue
          node.setNext(_current.getNext.getNext)
          _current.getNext.setNext(node)
          @count += 1
          return
        else
          _current = _current.getNext
        end
      end
    end
  end

  def remove(node)
    if @head && @head.getValue == node.getValue
      @head = @head.getNext
      @count -= 1
    else
      _current = @head
      while _current && _current.getNext do
        if _current.getNext.getValue == node.getValue
          _current.setNext(_current.getNext.getNext)
          @count -= 1
        else
          _current = _current.getNext
        end
      end
    end
  end

  def pop
    _current = @head
    if _current && !_current.getNext
      @head = nil
      @count = 0
    else
      while _current && _current.getNext do
        unless _current.getNext.getNext
          _current.setNext(nil)
          @count -= 1
          return
        else
          _current = _current.getNext
        end
      end
    end
  end

end
